packages/oracle-node/src/fetchers/twap-cache-service/TwapCacheService.ts

```
@@ -0,0 +1,209 @@
+        } as unknown as CacheServiceResponse);
+        promises.push(promise);
+        continue;
+      }
+      const url = `${config.historicalDataPackagesUrl}/${this.dataServiceId}/${timestampForHistoricalRequest}`;
+      const promise = axios.get<DataPackageFromCacheResponse>(url);
+      promises.push(promise);
+    }
+    return await Promise.allSettled(promises);
+  }
+
+  override extractPrice(
+    twapDataFeedId: string,
+    responses: RequestIdToResponse<CacheServiceResponsePromise[]>
+  ): number | undefined {
+    const twapResponses = responses[twapDataFeedId];
+    const { dataFeedId } =
+      TwapCacheServiceFetcher.parseTwapAssetId(twapDataFeedId);
+
+    const dataPackagesToCalculateTwap: DataPackageFromCache[] = [];
+    for (const response of twapResponses) {
+      const dataPackageFromThisNode =
+        TwapCacheServiceFetcher.getDataPackageFromThisNode(
+          response,
+          dataFeedId
+        );
+      if (dataPackageFromThisNode) {
+        dataPackagesToCalculateTwap.push(dataPackageFromThisNode);
+      }
+    }
+    TwapCacheServiceFetcher.validateSignatures(dataPackagesToCalculateTwap);
+    storeDataPackagesInLocalCache(dataPackagesToCalculateTwap);
+    return TwapCacheServiceFetcher.getTwapValue(dataPackagesToCalculateTwap);
+  }
+
+  private static parseTwapAssetId(twapDataFeedId: string): {
+    dataFeedId: string;
+    minutesOffset: number;
+  } {
+    const chunks = twapDataFeedId.split("-");
+    return {
+      dataFeedId: chunks[0],
+      minutesOffset: Number(chunks[chunks.length - 1]),
+    };
+  }
+
+  private static roundTimestampToMinute = (timestamp: number): number => {
+    return (
+      Math.floor(timestamp / ONE_MINUTE_IN_MILLISECONDS) *
+      ONE_MINUTE_IN_MILLISECONDS
+    );
+  };
+
+  private static getDataPackageFromThisNode(
+    response: CacheServiceResponsePromise,
+    dataFeedId: string
+  ): DataPackageFromCache | undefined {
```

How is it different to return undefined vs. throw from this method? I mean what are the use cases for both situations? It looks like we accept rejections (in a promise sense) from the server, but we throw on invalid data being returned.

---

If the promise failed for some reason, we can accept a small gap in TWAP calculation but if data packages are fetched but there isn't a data package signed by this node it means that we did something wrong and we should throw an error. 

But now I think it could lead to the situation that one node for some time was working incorrectly and throwing this error will make it impossible to calculate TWAP. I will remove this throw.